app/client/components/RegionFocusSwitcher.ts

```diff
@@ -0,0 +1,684 @@
+        // if we are not on a grist doc, whole page is always focusable
+        if (!gristDoc) {
+          return true;
+        }
+        // on a grist doc, panel content is focusable only if it's the current region
+        const current = use(this._state).region;
+        if (current?.type === 'panel' && current.id === id) {
+          return true;
+        }
+        // on a grist doc, main panel is focusable only if we are not the actual document view
+        if (id === "main") {
+          return this._canTabThroughMainRegion(use);
+        }
+        return false;
+      }),
+      cssFocusedPanel.cls('-focused', use => {
+        const current = use(this._state);
+        return current.initiator?.type === 'cycle' && current.region?.type === 'panel' && current.region.id === id;
+      }),
+    ];
+  }
+
+  private _cycle(direction: 'next' | 'prev') {
+    const gristDoc = this._getGristDoc();
+    const cycleRegions = getCycleRegions(gristDoc);
+    this.focusRegion(getSibling(
+      this._state.get().region,
+      cycleRegions,
+      direction,
+      gristDoc
+    ), {initiator: {type: 'cycle'}});
+  }
+
+  /**
+   * When clicking on a grist doc page:
+   *   - if necessary, make it easier to tab through things inside panels by "unfocusing" the view section,
+   *   - make sure the internal current region info is set when user clicks on the view layout.
+   */
+  private _onClick(event: MouseEvent) {
+    const gristDoc = this._getGristDoc();
+    if (!gristDoc) {
+      return;
+    }
+    this._commandsHistory = [];
+    const closestRegion = (event.target as HTMLElement)?.closest(`[${ATTRS.regionId}]`);
+    if (!closestRegion) {
+      return;
+    }
+    const targetRegionId = closestRegion.getAttribute(ATTRS.regionId);
+    const targetsMain = targetRegionId === 'main';
+    const current = this._state.get().region;
+    const currentlyInSection = current?.type === 'section';
+
+    if (targetsMain && !currentlyInSection) {
+      this.focusRegion(
+        {type: 'section', id: gristDoc.viewModel.activeSectionId()},
+        {initiator: {type: 'mouse', event}}
+      );
+      return;
+    }
+
+    const focusPanel = () => {
+      this.focusRegion(
+        {type: 'panel', id: targetRegionId as Panel},
+        {initiator: {type: 'mouse', event}}
+      );
+    };
+
+    // When not targeting the main panel, we don't always want to focus the given region _on click_.
+    // We only do it if clicking an empty area in the panel, or a focusable element like an input.
+    // Otherwise, we assume clicks are on elements like buttons or links,
+    // and we don't want to lose focus of current section in this case.
+    // For example I don't want to focus out current table if just click the "undo" button in the header.
+    if (!targetsMain && isFocusableElement(event.target)) {
+      focusPanel();
+    }
+
+    if (!targetsMain && getPanelElement(targetRegionId as Panel) === event.target) {
```

Comment:
Isn't `getPanelElement(targetRegionId as Panel)` the same as `closestRegion`?

Also, I think a slight refactoring would make this logic easier to follow. Something like this:
```
if (targetsMain) {
  if (!currentlyInSection) {
    this.focusRegion(...section...);
  }
} else {
  if (isFocusableElement(event.target) || closestRegion === event.target) {
    this.focusRegion(...panel...);
  }
}
```